src/streamlink/plugins/ustreamtv.py

```diff
@@ -1,355 +1,449 @@
 from random import randint
-from socket import error as SocketError
-from threading import Event, Thread
-from time import sleep
-from urllib.parse import unquote_plus, urljoin, urlparse, urlunparse
+from threading import Event, RLock
+from typing import Any, Callable, Deque, Dict, List, NamedTuple, Union
+from urllib.parse import urljoin, urlunparse
 
-import websocket
+from requests import Response
 
 from streamlink.exceptions import PluginError, StreamError
 from streamlink.plugin import Plugin, PluginArgument, PluginArguments, pluginmatcher
 from streamlink.plugin.api import useragents, validate
-from streamlink.stream.dash_manifest import sleep_until, utc
-from streamlink.stream.flvconcat import FLVTagConcat
-from streamlink.stream.segmented import (SegmentedStreamReader, SegmentedStreamWorker, SegmentedStreamWriter)
+from streamlink.plugin.api.websocket import WebsocketClient
+from streamlink.stream.ffmpegmux import MuxedStream
+from streamlink.stream.segmented import SegmentedStreamReader, SegmentedStreamWorker, SegmentedStreamWriter
 from streamlink.stream.stream import Stream
 from streamlink.utils.parse import parse_json
 
+
 log = logging.getLogger(__name__)
 
-Chunk = namedtuple("Chunk", "num url available_at")
-ChunkData = namedtuple("ChunkData", "chunk_id chunk_time hashes current_timestamp")
 
+# TODO: use dataclasses for stream formats after dropping py36 to be able to subclass
+class StreamFormatVideo(NamedTuple):
+    contentType: str
+    sourceStreamVersion: int
+    initUrl: str
+    segmentUrl: str
+    bitrate: int
+    height: int
+
+
+class StreamFormatAudio(NamedTuple):
+    contentType: str
+    sourceStreamVersion: int
+    initUrl: str
+    segmentUrl: str
+    bitrate: int
+    language: str
+
+
+class Segment(NamedTuple):
+    num: int
+    duration: int
+    available_at: datetime
+    hash: str
+    path: str
 
-class ModuleInfoNoStreams(Exception):
-    pass
+    # the segment URLs depend on the CDN and the chosen stream format and its segment template string
+    def url(self, base: str, template: str) -> str:
+        return urljoin(
+            base,
+            f"{self.path}/{template.replace('%', str(self.num), 1).replace('%', self.hash, 1)}"
+        )
 
 
-class UHSClient:
-    """
-    API Client, reverse engineered by observing the interactions
-    between the web browser and the ustream servers.
-    """
-    API_URL = "ws://r{0}-1-{1}-{2}-ws-{3}.ums.ustream.tv:1935/1/ustream"
-    APP_ID, APP_VERSION = 3, 2
-    api_schema = validate.Schema({
-        "args": [object],
-        "cmd": validate.text
+class UStreamTVWsClient(WebsocketClient):
+    API_URL = "wss://r{0}-1-{1}-{2}-ws-{3}.ums.ustream.tv:1935/1/ustream"
+    APP_ID = 3
+    APP_VERSION = 2
+
+    STREAM_OPENED_TIMEOUT = 6
+
+    _schema_cmd = validate.Schema({
+        "cmd": str,
+        "args": [{str: object}],
+    })
+    _schema_stream_formats = validate.Schema({
+        "streams": [validate.any(
```

Comment:
https://video.ibm.com/nasahdtv

```
[plugins.ustreamtv][error]
Unable to validate result: Unable to validate key 'streams':
Unable to validate key 'contentType': 'text/vtt' does not equal 'video/mp4'
or Unable to validate key 'contentType': 'text/vtt' does not equal 'audio/mp4'

[plugins.ustreamtv][debug] Processing 'moduleInfo' - 'stream'
error: No playable streams found on this URL: https://video.ibm.com/nasahdtv
```

Reply:
Yeah, I didn't check streams with subtitles. I'll take a look at this tomorrow. Shouldn't be too much work. One more validation schema, one more queue and stream for the muxed output, and no initialization segment.

Reply:
Do you have a stream with actual subtitles? The ones on the nasahdtv channel exist, but there's no/invalid content, so muxing the subtitle stream doesn't work. The `language` property is set to `unknown`, so maybe this needs to be filtered out.

This is the content of a WebVTT segment:
```vtt
WEBVTT


```

I've updated the branch with new changes, but without filtering out "invalid" subtitles